pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentUploadDownloadRestletResource.java

```diff
@@ -136,37 +136,30 @@ public Response downloadSegment(
           Response.Status.FORBIDDEN);
     }
 
-    FileUploadPathProvider provider;
-    try {
-      provider = new FileUploadPathProvider(_controllerConf);
-    } catch (Exception e) {
-      throw new ControllerApplicationException(LOGGER, e.getMessage(), Response.Status.INTERNAL_SERVER_ERROR, e);
-    }
-
+    segmentName = URIUtils.decode(segmentName);
+    URI dataDirURI = ControllerFilePathProvider.getInstance().getDataDirURI();
     Response.ResponseBuilder builder = Response.ok();
     File segmentFile;
     // If the segment file is local, just use it as the return entity; otherwise copy it from remote to local first.
-    if (CommonConstants.Segment.LOCAL_SEGMENT_SCHEME.equals(provider.getBaseDataDirURI().getScheme())) {
-      segmentFile =
-          new File(provider.getBaseDataDir(), StringUtil.join(File.separator, tableName, URIUtils.decode(segmentName)));
+    if (CommonConstants.Segment.LOCAL_SEGMENT_SCHEME.equals(dataDirURI.getScheme())) {
+      segmentFile = new File(new File(dataDirURI), StringUtil.join(File.separator, tableName, segmentName));
       if (!segmentFile.exists()) {
         throw new ControllerApplicationException(LOGGER,
             "Segment " + segmentName + " or table " + tableName + " not found in " + segmentFile.getAbsolutePath(),
             Response.Status.NOT_FOUND);
       }
       builder.entity(segmentFile);
     } else {
-      final URI segmentFileURI =
-          URIUtils.getUri(provider.getBaseDataDirURI().toString(), tableName, URIUtils.decode(segmentName));
-      PinotFS pinotFS = PinotFSFactory.create(provider.getBaseDataDirURI().getScheme());
-      if (!pinotFS.exists(segmentFileURI)) {
+      URI remoteSegmentFileURI = URIUtils.getUri(dataDirURI.toString(), tableName, URIUtils.encode(segmentName));
+      PinotFS pinotFS = PinotFSFactory.create(dataDirURI.getScheme());
+      if (!pinotFS.exists(remoteSegmentFileURI)) {
         throw new ControllerApplicationException(LOGGER,
-            "Segment " + segmentName + " or table " + tableName + " not found in " + segmentFileURI,
+            "Segment: " + segmentName + " of table: " + tableName + " not found at: " + remoteSegmentFileURI,
             Response.Status.NOT_FOUND);
       }
-      segmentFile = new File(StringUtil.join(File.separator, _controllerConf.getLocalTempDir(), tableName,
-          StringUtil.join("_", segmentName, String.valueOf(System.nanoTime()))));
-      pinotFS.copyToLocalFile(segmentFileURI, segmentFile);
+      segmentFile = new File(ControllerFilePathProvider.getInstance().getFileDownloadTempDir(),
+          StringUtil.join(File.separator, tableName, segmentName + "-" + System.nanoTime()));
```

Comment:
is uuid better than nanotime? i am fine leaving it as is

Reply:
Conceptually yes, but out of the scope of this pr, so leave it as is.